openedx/features/enterprise_support/enrollments/utils.py

```diff
@@ -0,0 +1,103 @@
+"""
+Utils for use in enrollment codebase such as views.
+"""
+import logging
+
+from django.core.exceptions import ObjectDoesNotExist  # lint-amnesty, pylint: disable=wrong-import-order
+from django.db import transaction
+
+from common.djangoapps.student.models import User
+from openedx.core.djangoapps.course_groups.cohorts import CourseUserGroup
+from openedx.core.djangoapps.enrollments import api
+from openedx.core.djangoapps.enrollments.errors import CourseEnrollmentError, CourseEnrollmentExistsError
+from openedx.core.djangoapps.enrollments.utils import add_user_to_course_cohort
+from openedx.core.lib.log_utils import audit_log
+from openedx.features.enterprise_support.enrollments.exceptions import (
+    CourseIdMissingException,
+    UserDoesNotExistException
+)
+
+log = logging.getLogger(__name__)
+
+
+def lms_enroll_user_in_course(username, course_id, mode,
+                              cohort_name=None,
+                              is_active=True,
+                              ):
+    """
+    Enrollment function meant to be called by edx-enterprise to replace the
+    current uses of the EnrollmentApiClient
+    The REST enrollment endpoint may also eventually also want to eventually reuse this function
+    since it's a subset of what the endpoint handles
+
+    Unlike the REST endpoint, this function does not check for enterprise enabled, or user api key
+    permissions etc. Those concerns are still going to be used by REST endpoint but this function
+    is meant for use from within edx-enterprise hence already presume such privileges.
+
+    Arguments:
+     - username (str): User name
+     - course_id (obj) : Course key obtained using CourseKey.from_string(course_id_input)
+     - mode (CourseMode): course mode
+     - cohort_name (str): Optional. If provided, user will be added to cohort
+     - is_active (bool): Optional. A Boolean value that indicates whether the
+        enrollment is to be set to inactive (if False). Usually we want a True if enrolling anew.
+
+    Returns: A serializable dictionary of the new course enrollment.
+    """
+    user = _validate_enrollment_inputs(username, course_id)
+
+    with transaction.atomic():
+        try:
+            # Will reactivate inactive enrollments.
+            response = api.add_enrollment(
+                username,
+                str(course_id),
+                mode=mode,
+                is_active=is_active,
+                enrollment_attributes=None
+            )
+            add_user_to_course_cohort(cohort_name, course_id, user)
+            log.info('The user [%s] has been enrolled in course run [%s].', username, course_id)
+            return response
+        except CourseEnrollmentExistsError as error:
```

Comment:
so with the comment just below try ("# Will reactivate inactive enrollments"), it sounds like it would perform a `CREATE OR UPDATE`, so why would we throw an exception? (I know you're just catching it here and not modifying the api itself, but I'm confused by how it would update existing, but inactive, enrollments if it throws an error when they exist)

Reply:
oh sorry did not get it, did you mean this particular exceptoin? CourseEnrollmentExistsError. The only exception I do not want to cause failures under is this one. Everything else should raise... Did that get to your question? if an enrollment already exists, we basically just log and move on

Reply:
I am slso not sure if the ` enrollment = _data_api().create_course_enrollment(username, course_id, mode, is_active, enterprise_uuid) ` is a create_or_update. I can check on that but it looks like it's just a create and will throw this enrollmentExistsError if one exists... and I saw this being caught in my logs

Reply:
that comment is confusing to me though.. when it says will activate inactive. I am guessing there is a case where it will make the enrollment active again (if it's in inactive state, meaning an update now), but then won't throw?

Reply:
after a slack thread with Adrienne and Brian I guess I will just remove that conditional and keep it! so... no change of logging from this fcn vs the view

Reply:
for ref: https://github.com/edx/edx-platform/pull/8698/files#r33624496 is where this audit_log was added a long long time ago I do not want to change how we report in case someone is auditing logs .. ( ecom maybe )

Reply:
Sorry. This exception (CourseEnrollmentExistsError) raised here, by name, implies that the add user or add enrollment call fails because the enrollment already exists.

But on line 52, you have a comment in this pr stating `# Will reactivate inactive enrollments.`

This makes it seem like if an enrollment exists, but is inactive, we'd throw and catch an error rather than actually update the status of it to "active". That is why I am confused here.

Reply:
```after a slack thread with Adrienne and Brian I guess I will just remove that conditional and keep it! so... no change of logging from this fcn vs the view```

I think this was meant to respond to a different comment I had around the audit_log call

Reply:
I realize maybe one off comments aren't easily visible in the github notifications, but my concern about the intention of this method (given the conflict between comment on line 52 and this particular exception) still applies.

Reply:
yeah so removed this comment because the docstring of api_enrollment already clarifies what it does, and it only sets the value to active True or False after creating.. does not update in-place as it seemed to be saying